### PR TITLE
feat: implement the enforcement for new invitation page

### PR DESCRIPTION
**Pull request by HGZ-20**
Implement the enforcement for new invitation page

Fix: https://github.com/casbin/casdoor/issues/2534

https://github.com/casbin/casdoor/assets/46803939/6e9f8a2c-6dac-4aa1-8f48-9ce9f25af8b0

### CODE REVIEW

**Comment by HGZ-20** [comment]
The modifications regarding the return status will be updated together based on the feedback received during the code review.

File: object/invitation.go [hunk 1]
```diff
@@ -99,6 +101,10 @@ func UpdateInvitation(id string, invitation *Invitation) (bool, error) {
 		return false, nil
 	}
 
+	if _, err := regexp.Compile(invitation.Code); err != nil {
```
Comment (hsluoyz) [inline object/invitation.go:104]: Define util function: `util.IsValidRegex(s string)`

File: object/invitation.go [hunk 2]
```diff
@@ -108,6 +114,10 @@ func UpdateInvitation(id string, invitation *Invitation) (bool, error) {
 }
 
 func AddInvitation(invitation *Invitation) (bool, error) {
+	if _, err := regexp.Compile(invitation.Code); err != nil {
```
Comment (hsluoyz) [inline object/invitation.go:117]: Define util function: `util.IsValidRegex(s string)`

File: object/user.go [hunk 3]
```diff
@@ -183,6 +183,7 @@ type User struct {
 	MfaPhoneEnabled     bool                  `json:"mfaPhoneEnabled"`
 	MfaEmailEnabled     bool                  `json:"mfaEmailEnabled"`
 	MultiFactorAuths    []*MfaProps           `xorm:"-" json:"multiFactorAuths,omitempty"`
+	InvitationFromCode  string                `xorm:"varchar(100) index" json:"invitationFromCode"`
```
Comment (hsluoyz) [inline object/user.go:186]: InvitationCode, remove "from"

File: object/user.go [hunk 4]
```diff
@@ -496,6 +497,24 @@ func GetUserByUserIdOnly(userId string) (*User, error) {
 	}
 }
 
+func GetUserByInvitationFromCode(owner string, invitationFromCode string) (*User, error) {
```
Comment (hsluoyz) [inline object/user.go:500]: GetUserByInvitationCode

File: object/user.go [hunk 5]
```diff
@@ -496,6 +497,24 @@ func GetUserByUserIdOnly(userId string) (*User, error) {
...
+	if owner == "" || invitationFromCode == "" {
+		return nil, nil
+	}
+
+	user := User{Owner: owner, InvitationFromCode: invitationFromCode}
```
Comment (hsluoyz) [inline object/user.go:505]: No "from"

File: object/user.go [hunk 6]
```diff
@@ -183,6 +183,7 @@ type User struct {
 	MfaPhoneEnabled     bool                  `json:"mfaPhoneEnabled"`
 	MfaEmailEnabled     bool                  `json:"mfaEmailEnabled"`
 	MultiFactorAuths    []*MfaProps           `xorm:"-" json:"multiFactorAuths,omitempty"`
```
Comment (hsluoyz) [inline object/user.go:185]: Add `Invitation string `xorm:"varchar(100) index" json:"invitation"`

Also record the invitation's name in User object

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/invitation.go [hunk 7]
```diff
@@ -132,3 +142,41 @@ func (invitation *Invitation) GetId() string {
...
+		reCode = reCode + "$"
+	}
+
+	re := regexp.MustCompile(reCode)
+	if !re.MatchString(invitationCode) {
+		return false
+	}
+
+	// Determine whether the invitation code is in the form of a regular expression other than pure numbers and letters
+	if match, err := regexp.MatchString("[^a-zA-Z0-9]", invitation.Code); err != nil {
```
[the remark this answers is not in the document]
Comment (HGZ-20) [reply object/invitation.go]: Use `regexp.QuoteMeta(s)` to replace

File: object/invitation.go (hunk 2 above)
Comment (HGZ-20) [reply object/invitation.go]: Updated to the following helper functions
```go
func IsRegexp(s string) (bool, error) {
	if _, err := regexp.Compile(s); err != nil {
		return false, err
	}
	return regexp.QuoteMeta(s) != s, nil
}
```

**Comment by HGZ-20** [comment]
demo1

https://github.com/casbin/casdoor/assets/46803939/b4592ece-4b68-4552-98f3-1a55f3500807

demo2

https://github.com/casbin/casdoor/assets/46803939/a1bdf577-65f0-4029-88f6-18197adb50d2

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**